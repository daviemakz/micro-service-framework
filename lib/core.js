'use strict';

// This file contains all the core functions which allow the eXchange server to
// do its work...

// Load Modules
var Notification = require('./../../open-daemon/core/util/logging.js');
var Network = require('./clients/common.js');
var Settings = require('./util/config.js');

// Load Templates
var CommandBodyObject = require('./template/command.js');
var ResponseBodyObject = require('./template/response.js');

// Invoke Configuration
var Configuration = new Settings();

// Load Configuration Settings
var MicroServerConnAttempts =
  Configuration.ServerCore.Settings.Connection.MicroAttempts;

// Invoke Server Objects
var DaemonMConnection = {
  Socket: '',
  Status: 0,
};
var DataMConnection = {
  Socket: '',
  Status: 0,
};
var NetworkMConnection = {
  Socket: '',
  Status: 0,
};
var OutreachMConnection = {
  Socket: '',
  Status: 0,
};
var SocialMConnection = {
  Socket: '',
  Status: 0,
};

//////////////// INVOKE SERVER CONNECTIONS /////////////////

// [Daemon] Micro Server
var StartMDaemon = function() {
  // Attempt Connection
  Network.InitiateMicroServerConnection(
    Configuration.ServerCore.Settings.XDaemon.PortListen,
    function(ConObject) {
      // Show Status
      if (ConObject.hasOwnProperty('error')) {
        console.log(
          'ERROR: Unable To Connect To Micro Server [Daemon]. Retrying...'
        );
        DaemonMConnection.Status = 0;
        StartMDaemon();
      } else {
        console.log(
          'NOTICE: Connected To Micro Server [Daemon]. Ready For Client Connections!'
        );
        DaemonMConnection.Status = 1;
      }
      // Store Socket Object
      DaemonMConnection.Socket = ConObject;
    }
  );
};

// [Data] Micro Server
var StartMData = function() {
  // Attempt Connection
  Network.InitiateMicroServerConnection(
    Configuration.ServerCore.Settings.XData.PortListen,
    function(ConObject) {
      // Show Status
      if (ConObject.hasOwnProperty('error')) {
        console.log(
          'ERROR: Unable To Connect To Micro Server [Data]. Retrying...'
        );
        DataMConnection.Status = 0;
        StartMData();
      } else {
        console.log(
          'NOTICE: Connected To Micro Server [Data]. Ready For Client Connections!'
        );
        DataMConnection.Status = 1;
      }
      // Store Socket Object
      DataMConnection.Socket = ConObject;
    }
  );
};

// [Network] Micro Server
var StartMNetwork = function() {
  // Attempt Connection
  Network.InitiateMicroServerConnection(
    Configuration.ServerCore.Settings.XNetwork.PortListen,
    function(ConObject) {
      // Show Status
      if (ConObject.hasOwnProperty('error')) {
        console.log(
          'ERROR: Unable To Connect To Micro Server [Network]. Retrying...'
        );
        NetworkMConnection.Status = 0;
        StartMNetwork();
      } else {
        console.log(
          'NOTICE: Connected To Micro Server [Network]. Ready For Client Connections!'
        );
        NetworkMConnection.Status = 1;
      }
      // Store Socket Object
      NetworkMConnection.Socket = ConObject;
    }
  );
};

// [Outreach] Micro Server
var StartMOutreach = function() {
  // Attempt Connection
  Network.InitiateMicroServerConnection(
    Configuration.ServerCore.Settings.XOutreach.PortListen,
    function(ConObject) {
      // Show Status
      if (ConObject.hasOwnProperty('error')) {
        console.log(
          'ERROR: Unable To Connect To Micro Server [Outreach]. Retrying...'
        );
        OutreachMConnection.Status = 0;
        StartMOutreach();
      } else {
        console.log(
          'NOTICE: Connected To Micro Server [Outreach]. Ready For Client Connections!'
        );
        OutreachMConnection.Status = 1;
      }
      // Store Socket Object
      OutreachMConnection.Socket = ConObject;
    }
  );
};

// [Social] Micro Server
var StartMSocial = function() {
  // Attempt Connection
  Network.InitiateMicroServerConnection(
    Configuration.ServerCore.Settings.XSocial.PortListen,
    function(ConObject) {
      // Show Status
      if (ConObject.hasOwnProperty('error')) {
        console.log(
          'ERROR: Unable To Connect To Micro Server [Social]. Retrying...'
        );
        SocialMConnection.Status = 0;
        StartMSocial();
      } else {
        console.log(
          'NOTICE: Connected To Micro Server [Social]. Ready For Client Connections!'
        );
        SocialMConnection.Status = 1;
      }
      // Store Socket Object
      SocialMConnection.Socket = ConObject;
    }
  );
};

//////////////// INITIALIZE MICRO SERVER /////////////////

// Initialize Connection [Daemon]
StartMDaemon();

// Initialize Connection [Data]
StartMData();

// Initialize Connection [Network]
StartMNetwork();

// Initialize Connection [Outreach]
StartMOutreach();

// Initialize Connection [Social]
StartMSocial();

/////////////////////////////////////////////////////////////

// FUNCTION : Handles Error(s) In Communication/Messages
function ProcessComError(data, message, id) {
  // Invoke Template(s)
  var ResObject = new ResponseBodyObject();
  var ComObject = new CommandBodyObject();

  // Get Parameters
  var CData = data;
  var ConId = id;
  var CMessage = message;

  // Check Empty Data
  if (!CData) {
    // Create Response Object
    var ResponseObject = ResObject;

    ////////////////
    // Build Response Object [Status - Transport]
    ResponseObject.Status.Transport = {
      Code: 2001,
      Message: 'No Data Recieved',
    };
    // Build Response Object [Status - Transport]
    ResponseObject.Status.Command = {
      Code: 200,
      Message: 'Command Not Executed - Transport Failiure / No Data Recieved',
    };
    // Build Response Object [ResBody - Error Details]
    ResponseObject.ResultBody.ErrData = {
      Entity: 'Main eXChange Server',
      Action: 'Message Error Handling',
      EType: 'ERROR',
      OriginalData: CData,
    };
    ////////////////

    // Send Response
    CMessage.reply(ResponseObject);

    // Notification
    Notification.LogEvent(
      null,
      null,
      null,
      'Main eXChange Server',
      'Message Error Handling',
      'ERROR',
      'No Data Recieved. MORE INFO: ' + ResponseObject,
      null,
      true
    );
  }

  // Return
  return;
}

// FUNCTION : Process Communication Request
function ProcessComRequest(data, message, id) {
  // Invoke Template(s)
  var ResObject = new ResponseBodyObject();
  var ComObject = new CommandBodyObject();

  // Get Parameters
  var CData = data;
  var ConId = id;
  var CMessage = message;

  // Micro Server Connection Attempt Count...
  var ConnectionAttempts = 0;

  var CheckConnection = function(
    recDataC,
    foreignSocketC,
    localSocketC,
    ConId
  ) {
    // Perform Action
    var ReturnResult = MicroServerCommunication(
      recDataC,
      foreignSocketC,
      localSocketC
    );

    // Check Connection, Execute Or Timeout...
    if (ReturnResult == 20) {
      if (ConnectionAttempts > MicroServerConnAttempts) {
        // Notification
        Notification.LogEvent(
          null,
          null,
          null,
          'Main eXChange Server',
          'Initiating Micro Server Connection',
          'ERROR',
          'Micro Server Connection Initiation Attempts - Maximum Reached',
          null,
          true
        );

        // Create Response Object
        var ResponseObject = ResObject;

        /////////////////////////
        // Build Response Object [Status - Transport]
        ResponseObject.Status.Transport = {
          Code: 2002,
          Message:
            'Micro Server Connection Initiation Attempts - Maximum Reached',
        };
        // Build Response Object [Status - Transport]
        ResponseObject.Status.Command = {
          Code: 201,
          Message: 'Command Not Executed - Transport Failiure',
        };
        // Build Response Object [ResBody - Error Details]
        ResponseObject.ResultBody.ErrData = {
          Entity: 'Main eXChange Server',
          Action: 'Micro Server Redirection',
          EType: 'ERROR',
          OriginalData: CData,
        };
        /////////////////////////

        // Send Response Back To Source
        foreignSocketC.reply(ResponseObject);

        // Close Socket
        foreignSocketC.conn.destroy();

        // Return
        return;
      } else {
        // Increment Connection Attempts
        ConnectionAttempts++;

        // Wait & Try Again...
        setTimeout(function() {
          CheckConnection(recDataC, foreignSocketC, localSocketC, ConId);
        }, 10);
      }
    } else {
      // CL
      console.log(
        '[' + ConId + '] Local Socket Connection Handed Over Successfully!'
      );

      // Return
      return;
    }
  };

  // Process Request [Redirection]
  if (CData.MSGDest == 'D-DAEMON') {
    // Check Connection & Send Data
    CheckConnection(CData, CMessage, DaemonMConnection, ConId);
  } else if (CData.MSGDest == 'D-DATA') {
    // Check Connection & Send Data
    CheckConnection(CData, CMessage, DataMConnection, ConId);
  } else if (CData.MSGDest == 'D-NETWORK') {
    // Check Connection & Send Data
    CheckConnection(CData, CMessage, NetworkMConnection, ConId);
  } else if (CData.MSGDest == 'D-OUTREACH') {
    // Check Connection & Send Data
    CheckConnection(CData, CMessage, OutreachMConnection, ConId);
  } else if (CData.MSGDest == 'D-SOCIAL') {
    // Check Connection & Send Data
    CheckConnection(CData, CMessage, SocialMConnection, ConId);
  } else {
    // Notification
    Notification.LogEvent(
      null,
      null,
      null,
      'Main eXChange Server',
      'Micro Server Message Routing',
      'ERROR',
      'Request Recieved But Destination Unknown. MORE INFO: ' + CData.MSGDest,
      null,
      true
    );

    // Create Response Object
    var ResponseObject = ResObject;

    /////////////////////////
    // Build Response Object [Status - Transport]
    ResponseObject.Status.Transport = {
      Code: 2005,
      Message: 'Request Recieved But Destination Unknown',
    };
    // Build Response Object [Status - Transport]
    ResponseObject.Status.Command = {
      Code: 201,
      Message: 'Command Not Executed - Transport Failiure',
    };
    // Build Response Object [ResBody - Error Details]
    ResponseObject.ResultBody.ErrData = {
      Entity: 'Main eXChange Server',
      Action: 'Micro Server Redirection',
      EType: 'ERROR',
      OriginalData: CData,
    };
    /////////////////////////

    // Reply
    CMessage.reply(ResponseObject);

    // Close Socket
    CMessage.conn.destroy();
  }

  // Return
  return;
}

// FUNCTION: Communicate With MicroServer
function MicroServerCommunication(recData, foreignSocket, localSocket) {
  // Get Parameters
  var RecievedData = recData;
  var FSocket = foreignSocket;
  var LSocket = localSocket;

  // Check Socket Readiness...
  if (LSocket.Status == 0) {
    // Return Fail (Local Socket Not Ready)
    return 20;
  } else {
    // Send Message To Micro Server
    LSocket.Socket.request('MICRO_COM_REQUEST', RecievedData, function(dataR) {
      // Send Micro Service Response To Source
      FSocket.reply(dataR);
      // Close Socket If Keep Alive Not Set
      if (RecievedData.MSGKeepAlive === false) {
        FSocket.conn.destroy();
      }
      return 10;
    });
  }
}

// EXPORTS
module.exports.ProcessComRequest = ProcessComRequest;
module.exports.ProcessComError = ProcessComError;
