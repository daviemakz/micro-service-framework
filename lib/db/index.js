'use strict';

// Load NPM modules
const db = require('quick.db');

// Define class
class FrameworkDatabase {
  constructor(options) {
    this.db = new db.table(options.databaseName);
    return this;
  }
}

module.exports = FrameworkDatabase;
