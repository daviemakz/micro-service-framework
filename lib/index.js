'use strict';

// Load NPM Modules
const getFreePort = require('find-free-port');
const makeDirectory = require('mkdirp');
const fs = require('fs');
const { exec } = require('child_process');
const { dirname } = require('path');

// Load Templates
const ResponseBodyObject = require('./template/response');

// Load libs
const MSCommon = require('./common');

// Load class
class ServiceCore extends MSCommon {
  // Constructor
  constructor(options) {
    // Allow access to 'this'
    super(options);
    // Invoke server tracking object
    this.microServerData = {};
    // Bind methods
    return (
      [
        'addServerToTracking',
        'initService',
        'initConnectionToService',
        'processComError',
        'microServerCommunication',
        'checkConnection',
        'processComRequest',
        'initiateMicroServerConnection',
      ].forEach(func => (this[func] = this[func].bind(this))) || this
    );
  }

  // FUNCTION: Add new server to tracking object. This contains all the information for microservices
  addServerToTracking(name, port) {
    // Check that the server doesnt already exist
    if (this.microServerData.hasOwnProperty(name)) {
      return false;
    }
    // Add the server to the tracking
    return (
      (this.microServerData[name] = {
        socket: void 0,
        status: false,
        error: false,
        port,
        funcs: {},
        process: void 0,
      }) && true
    );
  }

  // FUNCTION: Initiate a connection to a microservice
  initService(name, callback) {
    // Find a free port
    return getFreePort(this.settings.portRangeStart, this.settings.portRangeFinish, (err, port) => {
      // Check that the retrieving a port was successful
      if (!err) {
        // Add service to tracking
        this.addServerToTracking(name, port);
        // Build micro service wrapper
        const microServiceWrapper = () => {
          // Reset error status
          this.microServerData[name].error = false;
          // Assign process to instance
          this.microServerData[name].process = exec(
            `node ${__dirname}/server/index.js`,
            {
              maxBuffer: 1024 * this.settings.maxBuffer,
              env: {
                parentPid: process.pid,
                name,
                port,
                operations: this.microServerInfo[name],
                settings: JSON.stringify(this.settings),
              },
            },
            (error, stdout, stderr) => {
              // Reset error status
              if (error || stderr) {
                this.microServerData[name].error = true;
              }
              // Show log
              handleStdOnData(name, 'event', `Micro service - ${name}: Process has exited!`);
            }
          );
        };
        // Process text
        const processStdio = (name, type, data) =>
          (
            `[Child process: ${type}] Micro service - ${name}: ${
              typeof data === 'object' ? JSON.stringify(data, null, 2) : data
            }` || ''
          ).trim();
        // Handle stdio
        const handleStdOnData = (name, type, data) => {
          // Build text
          const logOutput = processStdio(`${name}/port:${port}`, type, data);
          // Write to log
          this.writeToLogFile(logOutput);
          // Show in parent console
          this.log(logOutput, 'log');
        };
        // Service exit handler assigner
        const assignEventHandlers = () => {
          // Assign to standard streams
          ['stdout', 'stderr'].forEach(event =>
            this.microServerData[name].process[event].on('data', data => handleStdOnData(name, event, data))
          );
          // onExit
          return ['exit'].forEach(event => this.microServerData[name].process.on(event, restartService));
        };
        // Service starter wrapper
        const startServiceProcess = () => {
          // Initialise service
          microServiceWrapper();
          // Assign process event handler
          assignEventHandlers();
          // Return nothing
          return void 0;
        };
        // Service restarter wrapper
        const restartService = () => setTimeout(startServiceProcess, this.settings.restartTimeout);
        // Start service initially
        startServiceProcess();
        // Initialise connection to the service
        this.initConnectionToService(name, port, (...args) => callback(...args));
      } else {
        callback(err, null);
      }
    });
  }

  // FUNCTION: Write output to log file
  writeToLogFile(contents) {
    return (
      this.settings.logPath &&
      makeDirectory(dirname(path), err => {
        // Throw error if failed to write to log file
        err && this.log(`Unable to write to log file. MORE INFO: ${err}`, 'warn');
        // Write the file
        fs.writeFile(this.settings.logPath, contents, () => void 0);
      })
    );
  }

  // FUNCTION: Connect To Server & Return Object (API gateway only!)
  initiateMicroServerConnection(port, callback) {
    // Get Variables
    let _connectionAttempts = 0;
    // Invoke Network Interface
    const _portSpeaker = this.invokeSpeakerPersistent(port);
    // Check Socket Is Ready & Execute
    const startMicroServiceConnection = () => {
      // Check If Socket Initialized Then Continue...
      if (Object.values(_portSpeaker.sockets).length === 0) {
        // Wait & Retry (including timeout)
        if (_connectionAttempts <= this.settings.microServiceConnectionTimeout) {
          // Try Again...
          setTimeout(() => {
            startMicroServiceConnection();
            return _connectionAttempts++;
          }, 10);
        } else {
          // Return Error Object
          _portSpeaker.error = 'Socket initialization timeout...';
          // Notification
          this.log(`Socket initialization timeout. PORT: ${portMicroServer}`, 'log');
        }
      } else {
        // Send Data To Destination
        this.log(`Service core successfully initialized socket on port: ${port}`, 'log');
        // Return Object Speaker
        return callback(_portSpeaker);
      }
    };
    // Check Connection & Start
    return startMicroServiceConnection();
  }

  // FUNCTION: Initiate a connection to a microservice
  initConnectionToService(name, port, callback) {
    // Initiate micro service connection
    return this.initiateMicroServerConnection(port, socket => {
      // Show status
      if (socket.hasOwnProperty('error')) {
        this.log('Unable to connect to service - ${name}. Retrying...', 'log');
        // Set status
        this.microServerData[name].status = false;
        // Retry
        return setTimeout(() => initConnectionToService(name, port, callback), this.settings.connectionTimeout);
      }
      this.log('Connected to service, ready for client connections!');
      // Set status
      this.microServerData[name].status = true;
      // Store Socket Object
      this.microServerData[name].socket = socket;
      // Callback
      return callback(true, socket);
    });
  }

  // FUNCTION : Handles Error(s) In Communication/Messages
  processComError(data, message, id) {
    // Get Parameters
    const _data = data;
    const _foreignSocket = message;
    // Check Empty Data
    if (!_data) {
      // Create Response Object
      const responseObject = new ResponseBodyObject();
      // Build Response Object [status - transport]
      responseObject.status.transport = {
        code: 2001,
        message: 'No data recieved',
      };
      // Build Response Object [status - transport]
      responseObject.status.command = {
        code: 200,
        message: 'Command not executed, tansport failure  or no data recieved!',
      };
      // Build Response Object [ResBody - Error Details]
      responseObject.resultBody.errData = {
        entity: 'Service core',
        action: 'Request error handling',
        errorType: 'ERROR',
        originalData: _data,
      };
      // Notification
      this.log(`No data received. MORE INFO: ${responseObject}`, 'log');
      // Send Response
      return _foreignSocket.reply(responseObject);
    }
    // Return
    return void 0;
  }

  // FUNCTION: Communicate With MicroServer
  microServerCommunication(recData, foreignSocket, localSocket, conId) {
    // Check Socket Readiness...
    return localSocket.status === 0
      ? 'connectionNotReady'
      : localSocket.socket.request('MICRO_COM_REQUEST', recData, res => {
          // Send Micro Service Response To Source
          foreignSocket.reply(res);
          // Close Socket If Keep Alive Not Set
          recData.keepAlive === false && foreignSocket.conn.destroy();
          // Show message in console
          recData.keepAlive === false
            ? this.log(`[${conId}] Service core has closed the connection!`, 'log')
            : this.log(
                `[${conId}] Service core has not closed this connection, this socket can be reused or manually closed via socket.conn.destroy()`,
                'log'
              );
          // Return
          return 'connectionReady';
        });
  }

  // Check connection
  checkConnection(recData, foreignSock, localSock, conId, conAttempts) {
    // Perform action
    const _connectionInstance = this.microServerCommunication(recData, foreignSock, localSock, conId);
    // Check Connection, Execute Or Timeout...
    if (_connectionInstance === 'connectionNotReady') {
      if (conAttempts > this.settings.microServiceConnectionAttempts) {
        // Notification
        this.log('Service connection initiation attempts, maximum reached');
        // Create Response Object
        const responseObject = new ResponseBodyObject();
        // Build Response Object [status - transport]
        responseObject.status.transport = {
          code: 2002,
          message: 'Service connection initiation attempts, maximum reached',
        };
        // Build Response Object [status - transport]
        responseObject.status.command = {
          code: 200,
          message: 'Command not executed, tansport failure!',
        };
        // Build Response Object [ResBody - Error Details]
        responseObject.resultBody.errData = {
          entity: 'Service core',
          action: 'Service redirection',
          errorType: 'ERROR',
          originalData: _data,
        };
        // Send Response Back To Source
        foreignSock.reply(responseObject);
        // Close Socket
        return foreignSock.conn.destroy();
      }
      // Increment Connection Attempts
      conAttempts++;
      // Wait & Try Again...
      setTimeout(() => this.checkConnection(recData, foreignSock, localSock, conId, conAttempts), 10);
    } else {
      // Console log
      return this.log(`[${conId}] Local socket connection handed over successfully!`);
    }
  }

  // FUNCTION : Process Communication Request
  processComRequest(data, message, id) {
    // Service Connection Attempt Count...
    const conAttempts = 0;
    // Get Parameters
    const _data = data;
    const _connectionId = id;
    const _foreignSocket = message;
    // Process Request [Redirection]
    if (this.microServerData.hasOwnProperty(_data.destination)) {
      // Get server data
      const _localSocket = this.microServerData[_data.destination];
      // Check Connection & Send Data
      this.checkConnection(_data, _foreignSocket, _localSocket, _connectionId, conAttempts);
    } else {
      // Notification
      this.log(`Request received but destination unknown. MORE INFO: ${_data.destination}`);
      // Create Response Object
      const responseObject = new ResponseBodyObject();
      // Build Response Object [status - transport]
      responseObject.status.transport = {
        code: 2005,
        message: 'Request recieved but destination unknown!',
      };
      // Build Response Object [status - transport]
      responseObject.status.command = {
        code: 200,
        message: 'Command not executed, transport failure!',
      };
      // Build Response Object [ResBody - Error Details]
      responseObject.resultBody.errData = {
        entity: 'Service core',
        action: 'Service redirection',
        errorType: 'ERROR',
        originalData: _data,
      };
      // Reply
      _foreignSocket.reply(responseObject);
      // Close Socket
      return _foreignSocket.conn.destroy();
    }
  }
}

// EXPORTS
module.exports = ServiceCore;
