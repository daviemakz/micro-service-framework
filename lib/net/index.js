'use strict';

// Load libs
const Listener = require('./listener');
const Speaker = require('./speaker');
const SpeakerReconnector = require('./speakerReconnector');

// Export [createListener]
exports.createListener = function(address) {
  return new Listener(address);
};

// Export [createSpeaker]
exports.createSpeaker = function() {
  const addresses = arguments.length >= 1 ? [].slice.call(arguments, 0) : [];
  return new Speaker(addresses);
};

// Export [createSpeaker]
exports.createSpeakerReconnector = function() {
  const addresses = arguments.length >= 1 ? [].slice.call(arguments, 0) : [];
  return new SpeakerReconnector(addresses);
};
