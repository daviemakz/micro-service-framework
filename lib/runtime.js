'use strict';

// EVENTS: unhandledRejection
process.on('unhandledRejection', (reason, p) =>
  console.log('ERROR - Unhandled Rejection Detected. MORE INFO: ', reason)
);

// EVENTS: rejectionHandled
process.on('rejectionHandled', (reason, p) =>
  console.log('ERROR - Rejection Handled Detected. MORE INFO: ', reason)
);

// EVENTS: SIGINT
process.on('SIGINT', function() {
  // Show Message
  console.log();
  console.log('CTRL + C Event Detected...');
  console.log('Script Will Exit...');
  // Exit...
  process.exit();
});
