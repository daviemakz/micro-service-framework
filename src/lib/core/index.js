'use strict';

import 'regenerator-runtime';

// Get a unique array
export function uniqueArray(arrArg) {
  return arrArg.filter((elem, pos, arr) => arr.indexOf(elem) === pos);
}

// Get a random list of elements from array
export function getRandomElements(arr, count) {
  const arrClone = [...arr];
  return [...Array(count)].map(
    () => arrClone.splice(Math.floor(Math.random() * arrClone.length), 1)[0]
  );
}

// Start service instance wrapper
export function startService(serviceInfo, instances) {
  return this.startServices(serviceInfo, instances);
}

// Stop service instance wrapper
export function stopService(name, instances) {
  return new Promise((resolve, reject) => {
    // Get highest number of instances which can be shutdown
    const requestedInstances = Math.abs(instances);
    const actualInstances = this.serviceData[name].port.length;
    const instancesToTerminate =
      requestedInstances > actualInstances
        ? actualInstances
        : requestedInstances;
    // If instance count is 0 then just resolve
    if (instancesToTerminate === 0) {
      return resolve(void 0);
    }
    // Get ports to close
    const ports = getRandomElements(
      this.serviceData[name].port,
      instancesToTerminate
    );
    // Assign ports which are being shutdown
    process.env.exitedProcessPorts = uniqueArray(
      [].concat(process.env.exitedProcessPorts, ports)
    );
    (typeof process.env.exitedProcessPorts === 'string'
      ? process.env.exitedProcessPorts.split(',')
      : process.env.exitedProcessPorts
    )
      .map((port) => parseInt(port, 10))
      .filter((exitedPort) => typeof exitedPort === 'number');
    // Process indexes
    const processIndexes = ports.map((port) =>
      this.getProcessIndex(name, port)
    );
    // Process exiter
    try {
      processIndexes.forEach(
        async (index, elIndex) =>
          new Promise((resolve, reject) => {
            // Show message
            this.log(
              `Service core will send kill command to the service: ${name}/port:${ports[elIndex]}`,
              'log'
            );
            // Send kill process message
            this.serviceData[name].clientSocket[index].request(
              'SERVICE_KILL',
              void 0,
              (res) => {
                this.log(
                  `Service core has recieved acknowledgement of kill command from: ${name}/port:${ports[elIndex]}`,
                  'log'
                );
                if (res.status.command.code === 200) {
                  resolve(true);
                } else {
                  reject(Error(false));
                }
              }
            );
          })
      );
      // Resolve
      return resolve(void 0);
    } catch (e) {
      return reject(e);
    }
  });
}

// End the framework, turn it off
export function end({ sendSuccess }) {
  // Set the success
  sendSuccess({
    result: {
      error: null,
      details: {},
      isSuccess: true,
      message: 'Shutting down micro service framework.'
    }
  });
  // Kill process
  return setTimeout(() => process.exit(), 1000);
}

// Store persistent data
export function storage({ data, sendError, sendSuccess }) {
  const { table, method, args } = data.body;
  return setImmediate(() =>
    this.databaseOperation(table, method, args, (isSuccess, result, error) => {
      // Assign function result
      if (isSuccess) {
        return sendSuccess({
          result: {
            isSuccess,
            result,
            message: 'The operation completed successfully!'
          }
        });
      }
      return sendError({
        result: {
          isSuccess,
          result,
          error,
          message: 'The operation failed!'
        },
        code: 401,
        message:
          'Command executed but an error occurred while attempting storage operation.'
      });
    })
  );
}

// Change the number of running instances
export async function changeInstances({ data, sendError, sendSuccess }) {
  const { name, instances } = data.body;
  const baseResponse = {
    error: null,
    details: {}
  };

  // Check if the service has already been defined
  if (!Object.keys(this.serviceInfo).includes(name)) {
    sendError({
      result: Object.assign(baseResponse, {
        isSuccess: false,
        message: `Service "${name}" was not found!`
      })
    });
  } else if (typeof instances !== 'number' || instances === 0) {
    sendError({
      result: Object.assign(baseResponse, {
        isSuccess: false,
        message: '"instance" property must be a number which is not 0!'
      })
    });
  } else {
    // Base response object
    const previousPorts = [].concat(...this.serviceData[name].port);

    // Get result
    const result =
      instances > 0
        ? await startService.call(
            this,
            { [name]: this.serviceInfo[name] },
            instances
          )
        : await stopService.call(this, name, instances);

    // Get next ports
    const nextPorts = this.serviceData[name].port;

    // Assign response object
    if (typeof result === 'undefined') {
      sendSuccess({
        result: Object.assign(baseResponse, {
          isSuccess: true,
          message: 'Services were modified successfully!',
          details: {
            instances: nextPorts.length,
            name,
            previousPorts,
            nextPorts
          }
        })
      });
    } else {
      sendError({
        result: Object.assign(baseResponse, {
          isSuccess: false,
          message: 'Services modification failed!',
          details: {
            instances: nextPorts.length,
            name,
            previousPorts,
            nextPorts
          }
        }),
        code: 402,
        message:
          'Command executed but an error occurred while attempting to change instances'
      });
    }
  }
}

// Export request functions
export * from './request';

// Export functions to be consumed
export default {
  end,
  storage,
  changeInstances,
  uniqueArray,
  getRandomElements
};
