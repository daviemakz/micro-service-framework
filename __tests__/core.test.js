'use strict';

import { uniqueArray, getRandomElements } from '../tmp/lib/core';
import { makeRequest, host, httpPortOne } from './utils/helpers';

// Test suite
describe('tmp/lib/core', () => {
  describe('uniqueArray()', () => {
    test('to match snapshot', () => {
      expect(uniqueArray([1, 1, 1, 2, 3, 4, 5])).toMatchSnapshot();
    });
  });

  describe('getRandomElements()', () => {
    test('to match snapshot', () => {
      expect(
        Array.isArray(getRandomElements([1, 1, 1, 2, 3, 4, 5], 2))
      ).toMatchSnapshot();
    });
  });

  describe('run on start & core operations', () => {
    describe('can execute a custom core operation on service core', () => {
      let output;

      beforeAll((done) => {
        return makeRequest({
          host,
          method: 'get',
          port: httpPortOne,
          path: '/reflect-core-operation',
          callback: (data) => {
            output = data;
            done();
          }
        });
      });

      test('http status code is 200', () => {
        expect(output.status).toEqual(200);
      });

      test('service request status code is true', () => {
        expect(output.data.status).toEqual(true);
      });

      test('response data is expected', () => {
        expect(output.data.response).toEqual('Reflect service core!');
      });
    });

    describe('can execute a custom core operation on micro service', () => {
      let output;

      beforeAll((done) => {
        return makeRequest({
          host,
          method: 'get',
          port: httpPortOne,
          path: '/get-local-storage',
          callback: (data) => {
            output = data;
            done();
          }
        });
      });

      test('http status code is 200', () => {
        expect(output.status).toEqual(200);
      });

      test('service request status code is true', () => {
        expect(output.data.status).toEqual(true);
      });

      test('response data is expected', () => {
        expect(output.data.response).toEqual({ key: 'management' });
      });
    });

    describe('runOnStart function saves the number which can be read by this function', () => {
      let output;

      beforeAll((done) => {
        return makeRequest({
          host,
          method: 'get',
          port: httpPortOne,
          path: '/get-saved-number',
          callback: (data) => {
            output = data;
            done();
          }
        });
      });

      test('http status code is 200', () => {
        expect(output.status).toEqual(200);
      });

      test('service request status code is true', () => {
        expect(output.data.status).toEqual(true);
      });

      test('response data is expected', () => {
        expect(output.data.response).toEqual({
          isSuccess: true,
          message: 'The operation completed successfully!',
          result: 500
        });
      });
    });
  });

  describe('storage()', () => {
    describe('can set data via a "storageService" and retrieve it via "serviceCore"', () => {
      let output;

      beforeAll((done) => {
        const number = 81;
        makeRequest({
          host,
          method: 'post',
          body: { number },
          port: httpPortOne,
          path: '/save-powered-number',
          callback: (data) => {
            output = data;
            done();
          }
        });
      });

      [
        true,
        {
          isSuccess: true,
          result: 6561,
          message: 'The operation completed successfully!'
        }
      ].forEach((response, index) => {
        test('http status code is 200', () => {
          expect(output.data[index].command.code).toEqual(200);
        });

        test('service request status code is true', () => {
          expect(output.data[index].status).toEqual(true);
        });

        test('response data is expected', () => {
          expect(output.data[index].response).toEqual(response);
        });
      });
    });
  });

  describe('changeInstances()', () => {
    describe('can increase the number of the "instanceService" and get responses from all processes', () => {
      let output;

      beforeAll((done) => {
        const addInstances = 2;
        makeRequest({
          host,
          method: 'post',
          body: { addInstances },
          port: httpPortOne,
          path: '/change-instances',
          callback: (data) => {
            output = data;
            done();
          }
        });
      });

      describe('check that all responses succeeded', () => {
        Array.from(Array(6)).forEach((_, index) => {
          test(`http status code is 200 for process ${index + 1}`, () => {
            expect(output.data[index].command.code).toEqual(200);
          });

          test(`service request status code is true for process ${
            index + 1
          }`, () => {
            expect(output.data[index].status).toEqual(true);
          });
        });
      });

      describe('there are only three instances', () => {
        test('instance count is expected', async () => {
          return new Promise((resolve) => {
            const checkOutput = () => {
              if (output) {
                resolve();
              } else {
                setTimeout(() => {
                  checkOutput();
                }, 10);
              }
            };
            checkOutput();
          }).then(() => {
            const uniqueInstances = [
              ...new Set(
                output.data
                  .map((proc) => proc.transport.responseSource.instanceId)
                  .filter((instId) => instId)
              )
            ];
            expect(uniqueInstances.length).toEqual(3);
          });
        });
      });
    });
  });
});
