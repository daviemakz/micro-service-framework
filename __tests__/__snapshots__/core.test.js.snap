// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tmp/lib/core getRandomElements() to match snapshot 1`] = `true`;

exports[`tmp/lib/core uniqueArray() to match snapshot 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
